### ISSUE TITLE
PAC doesnt follow lag compensation in TTT

### ISSUE THREAD

**Issue by Earu** [CapsAdmin/pac3#1341]
To reproduce this issue, have a player wear an outfit with model parts parented to any bone of a player. Hit the player while they are moving, notice how the part are desync'd.

**Comment by homonovus** [comment]
related source links
https://github.com/ValveSoftware/source-sdk-2013/blob/0d8dceea4310fde5706b3ce1c70609d72a38efdf/sp/src/game/server/player_lagcompensation.cpp#L410
https://github.com/ValveSoftware/source-sdk-2013/blob/0d8dceea4310fde5706b3ce1c70609d72a38efdf/sp/src/game/server/player_lagcompensation.cpp#L611

**Comment by CapsAdmin** [comment]
related garrysmod issue
https://github.com/Facepunch/garrysmod-issues/issues/884

**Comment by wrefgtzweve** [comment]
If i understand the issue correctly the issue is when the player gets lagcompensated.
Did a test with this, seems like only `GetBonePosition` is affected while `GetBoneMatrix` + `GetTranslation` doesnt

https://github.com/CapsAdmin/pac3/assets/69946827/ace5e1d2-533a-4810-8638-2d61124c217f

The test code:
```lua
hook.Add( "HUDPaint", "BoneExample", function()
    for _, ply in ipairs( player.GetAll() ) do
        if not ply:Alive() then continue end
        if ply == LocalPlayer() then continue end

        local bone = ply:LookupBone( "ValveBiped.Bip01_Head1" )
        if not bone then continue end

        local matrix = ply:GetBoneMatrix( bone )
        if not matrix then continue end

        local pos = matrix:GetTranslation()
        local toScreen = pos:ToScreen()

        draw.SimpleText( "GetBoneMatrix", "DermaDefault", toScreen.x, toScreen.y, color_white, TEXT_ALIGN_CENTER, TEXT_ALIGN_CENTER )

        local pos = ply:GetBonePosition( bone )
        local toScreen = pos:ToScreen()

        draw.SimpleText( "GetBonePosition", "DermaDefault", toScreen.x, toScreen.y, color_white, TEXT_ALIGN_CENTER, TEXT_ALIGN_CENTER )
    end
end )
```

**Comment by TW1STaL1CKY** [comment]
Yeah, that is the problem. It looks like PAC does mainly use Entity:GetBoneMatrix, though there are a couple of spots still using Entity:GetBonePosition it seems. Maybe those few instances are the cause?

**Comment by CapsAdmin** [comment]
Last time I looked into this, I couldn't figure out how to use GetBoneMatrix without breaking everything.

**Comment by Earu** [comment]
> Last time I looked into this, I couldn't figure out how to use GetBoneMatrix without breaking the part placement of old outfits.

How about today? Maybe you have more ideas than back then? 😁

**Comment by TW1STaL1CKY** [comment]
Using GetBoneMatrix seems to work fine for me, I'm not sure how old an "old outfit" would be though, the outfits that I've tried have been fine

Before - pac.GetBonePosAng using GetBonePosition:
https://github.com/CapsAdmin/pac3/assets/1499739/d508cf27-92c8-41fb-9c71-8d5a5d119c94

After - pac.GetBonePosAng using GetBoneMatrix: (the nametags I have use GetBonePosition so you can kind of compare the difference using it)
https://github.com/CapsAdmin/pac3/assets/1499739/babe7b9e-f0d8-4b49-b01d-d4f59f3579b4

I can push/PR this to the develop branch or a new branch for further testing?

**Comment by TW1STaL1CKY** [comment]
I was told to just push it so I've pushed it to develop😁

**Comment by pingu7867** [comment]
Sorry for the delay, I didn't find out about it until now because I don't sync my local install. I had a tester reach out to me.

This breaks clips on any valid bone.
Yet static models or placement on invalidbone, as well as special pac bones, are fine. Not so with other bones.

https://github.com/user-attachments/assets/f06486fd-af99-42bc-a844-d2441dae0143

Changing this is a serious backward compatibility problem. People do use clips on character models. So until we make a serious attempt (one that doesn't involve micro-edits but rather a fundamental look at the drawtime and bone building algorithms) at solving the whole lagging behind issue (a closely related issue that was only "solved" by trying different combinations of translucent in your hierarchy), it should be an opt-in field in the base_movable or base_drawable class.


https://github.com/user-attachments/assets/8291368f-7f37-42f7-8be7-ec4afba4e7f3

**Comment by TW1STaL1CKY** [comment]
Only clip parts seem to have this issue. If they're used on a model part that parents to a bone on that model part, it lags behind for some reason. Calling SetupBones for the model before it applies the clip plane seems to resolve it.

I've made a fix in a PR you can check out before it gets merged into develop: #1391
